book-content/chapters/04-essential-types-and-annotations.md

```diff
@@ -0,0 +1,2366 @@
+
+const modifyUser = (user: User[], id: string, makeChange: MakeChangeFunc) => {
+  // function body
+};
+```
+
+Both techniques behave the same, but if you need to reuse the `makeChange` function type, a type alias is the way to go.
+
+#### Solution 5: Functions Returning `void`
+
+Let's start by annotating the `listener` parameter to be a function. For now, we'll specify that it returns a string:
+
+```ts
+const addClickEventListener = (listener: () => string) => {
+  document.addEventListener("click", listener);
+};
+```
+
+The problem is that we now have an error when we call `addClickEventListener` with a function that returns nothing:
+
+```ts twoslash
+// @errors: 2345
+const addClickEventListener = (listener: () => string) => {
+  document.addEventListener("click", listener);
+};
+
+// ---cut---
+addClickEventListener(() => {
+  console.log("Clicked!");
+});
+```
+
+The error message tells us that the `listener` function is returning `void`, which is not assignable to `string`.
+
+This suggests that instead of typing the `listener` parameter as a function that returns a string, we should type it as a function that returns `void`:
+
+```ts
+const addClickEventListener = (listener: () => void) => {
+  document.addEventListener("click", listener);
+};
+```
+
+This is a great way to tell TypeScript that we don't care about the return value of the `listener` function.
+
+#### Solution 6: `void` vs `undefined`
+
+The solution is to change the of `callback` to `() => void`:
+
+```ts
+const acceptsCallback = (callback: () => void) => {
+  callback();
+};
+```
+
+Now we can pass in `returnString` without any issues. This is because `returnString` returns a `string`, and `void` tells TypeScript to ignore the return value when comparing them.
+
+So if you really don't care about the result of a function, you should type it as `() => void`.
+
+#### Solution 7: Typing Async Functions
+
+You might be tempted to try passing a type argument to `fetch`, similar to how you would with `Map` or `Set`.
+
+However, hovering over `fetch`, we can see that it doesn't accept type arguments:
+
+```ts
+// @noErrors
+const response = fetch<number>("https://api.example.com/data");
+//               ^?
+```
+
+We also can't add a type annotation to `response.json()` because as it doesn't accept type arguments either:
+
+```ts twoslash
+// @errors: 2558
+const response = await fetch("https://api.example.com/data");
+
+// ---cut---
+const data: number = await response.json<number>();
+```
+
+One thing that will work is to specify that `data` is a `number`:
+
+```ts
+const response = await fetch("https://api.example.com/data");
+
+// ---cut---
+const data: number = await response.json();
+```
+
+This works because `data` was `any` before, and `await response.json()` returns `any`. So now we're putting `any` into a slot that requires a `number`.
+
+However, the best way to solve this problem is to add a return type to the function. In this case, it should be a `number`:
```

Comment:
why would it be the best way? arguably i'd say that it's better to annotate the variable (and return type optionally if that's ur thing). By annotating the variable u benefit from the annotation on the lines between ur `.json()` and the return statement.